[1단계 - JDBC 라이브러리 구현하기] - 엠제이(최민준) 미션 제출합니다

theminjunchoi opened the pull request:
안녕하세요 뽀스티
레벨1 첫 번째 미션과 레벨4 마지막 미션을 같이 하네요.
마음이 되게 몽글몽글뜨끈뜨근웅장웅장말랑말랑해지는 것 같아요.
<img width="218" height="231" alt="image" src="https://github.com/user-attachments/assets/dbee5eb6-f9f3-4f94-8da3-52c0adaf69d0" />

편하게 리뷰 부탁드립니다~!

theminjunchoi committed:
build: 테스트 동작을 위한 JUnit Platform Launcher 의존성 추가

theminjunchoi committed:
feat: JDBC 템플릿에 쿼리 및 업데이트 메서드 추가

- `queryForObject` 메서드 추가: 단일 결과 조회 지원
- `query` 메서드 추가: 다중 결과 조회 지원
- `update` 메서드 추가: 업데이트 쿼리 실행 지원
- `PreparedStatementCallback` 및 `RowMapper` 인터페이스 추가
- `EmptyResultDataAccessException` 및 `IncorrectResultSizeDataAccessException` 예외 클래스 추가

theminjunchoi committed:
refactor: UserDao에 JDBC 템플릿 적용

- JDBC 템플릿 기반으로 UserDao 리팩토링
- 반복적인 커넥션 및 리소스 처리 코드 제거
- `update` 및 `queryForObject` 메서드로 코드 간소화
- `RowMapper` 인터페이스로 매핑 로직 분리 및 재사용성 확보

theminjunchoi committed:
refactor: schema.sql 테이블 생성 및 삭제 로직 수정

- `user_history` 및 `users` 테이블 삭제 로직 추가
- `create table` 구문에서 `if not exists` 제거
- 테이블 정의 가독성 및 일관성 향상

ykmxxi commented on app/src/main/resources/schema.sql:
```diff
@@ -1,12 +1,15 @@
-create table if not exists users (
+drop table if exists user_history;
+drop table if exists users;
+
+create table users (
```
기존 schema 에서 변경하신 이유가 있을까요 엠줴이?

theminjunchoi replied:
`UserDaoTest`를 실행하다보면 `setup`에서 계속 gugu라는 유저를 넣어주더라고요.

이렇게 되면 테스트가 실행될 때마다 gugu라는 유저가 반복적으로 추가되는데,
`findByAccount` 테스트에서 gugu라는 계정이 여러개 검색이 되는 문제가 발생했었습니다.

해결 방법으로는 `AfterEach`로 gugu라는 유저를 제거해주거나,
지금처럼 테이블 자체를 제거 후 새롭게 만드는 방법이 있을 것 같은데,
후자 방법이 더 빠르게 문제를 해결할 수 있어서 이렇게 구현했어용

ykmxxi replied:
👍

ykmxxi commented on jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java:
```diff
@@ -14,4 +23,89 @@ public class JdbcTemplate {
+            throw new IncorrectResultSizeDataAccessException(1, results.size());
+        }
+        
+        return results.get(0);
```
```suggestion
        return results.getFirst();
```

컬렉션 프레임워크에서 제공하는 메서드를 써서 명확하게 보여줘도 좋을것 같아요

theminjunchoi replied:
인정

ykmxxi commented on jdbc/src/main/java/com/interface21/jdbc/core/PreparedStatementCallback.java:
```diff
@@ -0,0 +1,10 @@
+import java.sql.SQLException;
+
+@FunctionalInterface
+public interface PreparedStatementCallback<T> {
```
이 인터페이스는 왜 필요한걸까요?? 두 가지 궁금증이 있어요

1. `PreparedStatement`는 어떤 역할을 하는 클래스일까요??
2. 왜 인터페이스를 통해 `PreparedStatement`를 인자로 넘겨줘 실행해야 하나요?

theminjunchoi replied:
>  1. PreparedStatement는 어떤 역할을 하는 클래스일까요??

SQL을 미리 컴파일해서 파라미터만 바꿔가며 실행시킬 수 있는 책임을 갖고 있고,
타입도 안전하게 바인딩이 되는 걸로 알고 있습니다!

그리고 아래 상황에서 userName에 `"admin'`과 같은 민감한 정보도 자동으로 이스케이프 처리되는 걸로 알고 있습니다!
```java
String sql = "SELECT * FROM users WHERE name = ?";
pstmt.setString(1, userName); 
```

theminjunchoi replied:
> 2. 왜 인터페이스를 통해 PreparedStatement를 인자로 넘겨줘 실행해야 하나요?

오호 좋은 질문이에요! 
핵심 문제는 **PreparedStatement의 생명주기 관리**인 것 같아요.
이 친구는 안닫아주면 메모리 누수, DB 커넥션 고갈을 야기할 수 있는데요,
이 책임을 한 곳으로 모아주기 위해 위와 같은 방법으로 구현했어요.

그리고 처음 JdbcTemplate를 공부하면서 실제 원본 코드 중에서 PreparedStatementCallback이 제일 기억에 남았다보니까 자연스레 이렇게 분리를 해줬네용

ykmxxi replied:
콜백 패턴을 활용하는 것 좋네요 👍  역시 천상계

ykmxxi commented on jdbc/src/main/java/com/interface21/jdbc/core/PreparedStatementCallback.java:
```diff
@@ -0,0 +1,10 @@
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+@FunctionalInterface
```
`IncorrectResultSizeDataAccessException`처럼 간단하게 어떤 클래스인지 설명하는 주석을 추가해도 좋을것 같아요 👍

theminjunchoi replied:
어익쿠야 적어준다는 걸 까먹었네요!
추가해주었습니다

ykmxxi reviewed (changes requested):
안녕하세요 민준초이!! 👍 

<img width="376" height="205" alt="image" src="https://github.com/user-attachments/assets/700bf82b-815a-46a0-b7ee-2ac82cd7521d" />

레벨1 첫 번째 미션과 레벨4 마지막 미션을 같이 하네요.
저도 마음이 되게 몽글몽글뜨끈뜨근웅장웅장말랑말랑해지는 것 같아요.

역시 천상계 답게 참 잘하시네요! 몇가지 질문이 있어 Request changes 요청 드려요! 답변 해주시고 리뷰 요청 보내시면 바로 approve & merge 하겠습니다

theminjunchoi committed:
docs: 테스트 클래스 내 격리 수준별 읽기 현상 테이블 업데이트

- `Read Uncommitted`, `Read Committed`, `Repeatable Read`, `Serializable`의 읽기 현상 정의 추가
- 관련 테스트 메서드 주석에 테이블 내용 반영:
  - 더티 리드
  - 비반복적 리드
  - 판독 리드
- 가독성 및 정확성 향상 목적

theminjunchoi committed:
test: Stage2Test 트랜잭션 테스트 케이스 수정 및 검증 로직 개선

- `Stage2Test` 테스트 시나리오에서 예상 결과 업데이트
  - `hasSize()` 및 `containsExactly()` 검증 값 수정
- `saveFirstTransactionWithMandatory`, `saveFirstTransactionWithNotSupported` 메서드 트랜잭션 설정 변경
- 트랜잭션 동작 확인 테스트의 정확성 및 일관성 향상

theminjunchoi committed:
refactor: 단일 결과 반환 로직 수정

- `results.get(0)`에서 `results.getFirst()`로 반환 방식 변경
- 코드 가독성 및 메서드의 의도 표현력 강화

theminjunchoi committed:
docs: EmptyResultDataAccessException 클래스에 JavaDoc 추가

- 클래스 설명 주석 추가: 결과가 없을 때 발생하는 예외 정의
- 코드 가독성 및 이해도 향상

theminjunchoi reviewed (commented):
안녕하세요 강민햄
리뷰 감사합니다~!
즐거운 저녁되세요~!

ykmxxi commented on jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java:
```diff
@@ -14,4 +23,89 @@ public class JdbcTemplate {
+        });
+    }
+
+    private <T> T execute(String sql, PreparedStatementCallback<T> callback) {
+        Connection conn = null;
+        PreparedStatement pstmt = null;
+        
+        try {
+            conn = dataSource.getConnection();
+            pstmt = conn.prepareStatement(sql);
+            
+            log.debug("Executing SQL: {}", sql);
+            
+            return callback.doInPreparedStatement(pstmt);
+            
+        } catch (SQLException e) {
+            log.error("SQL execution failed: {}", sql, e);
+            throw new DataAccessException("SQL execution failed: " + sql, e);
+        } finally {
+            closeStatement(pstmt);
+            closeConnection(conn);
+        }
+    }
```
```suggestion
private <T> T execute(String sql, PreparedStatementCallback<T> callback) {
    try (Connection conn = dataSource.getConnection();
         PreparedStatement pstmt = conn.prepareStatement(sql)) {
        
        log.debug("Executing SQL: {}", sql);
        return callback.doInPreparedStatement(pstmt);
    } catch (SQLException e) {
        log.error("SQL execution failed: {}", sql, e);
        throw new DataAccessException("SQL execution failed: " + sql, e);
    }
}
```

이렇게 try-with-resources 구문을 활용하면 더 안전하게 리소스를 close하고 불필요한 close 메서드 구현을 할 필요가 없을것 같아요. 어떤 경우 `null`이 들어오는 것을 고려해 따로 close 메서드를 구현했을까요??

theminjunchoi replied:
@ykmxxi 
오 맞네요!

null 체크를 한 이유는 아래와 같아요.
dataSource.getConnection()에서 예외 발생 시 conn이 null
conn.prepareStatement(sql)에서 예외 발생 시 pstmt가 null

이런 경우에 finally 블록에서 null.close()를 호출하면 NPE가 발생하기 때문에 null 체크가 필요했습니다.
하지만 뽀스티 말처럼 try-with-resources를 사용하면 더 좋을 것 같네요!

살짝 고민되는 건 트랜잭션을 고려했을 때, 기존 트랜잭션이 존재 시에는 그때 사용하던 Connection을 재사용해야하고 메서드가 종료되어도 Connection이 닫히면 안돼서 그땐 try-with-resources를 사용하기 힘들 것 같은데 그건 그때가서 생각해보겠습니다!

뽀스티가 제안해준 방식으로 step2에 수정해볼게요!

ykmxxi reviewed (approved):
<img width="490" height="414" alt="image" src="https://github.com/user-attachments/assets/f48f0265-09f1-4928-a5d4-56a65e42472d" />

민준씌 역시 천상계 답게 좀 치시네요 👍 

코멘트 하나 남겼으니 확인해보시면 좋을것 같아요. 다음 스떼프로 넘어가시죠
다음 단계도 너무 기대되서 마음이 몽글몽글뜨끈뜨근웅장웅장말랑말랑해지는것 같네요
